url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/robot/SOCRobotDM.java

```diff
@@ -0,0 +1,2114 @@
+	    pending.push(new Pair(new NodeLenVis(j, len+1, newVis), dataPair));
+	    pathEnd = false;
+	  }
+	}	 
+	j = coord - 0x10;
+	edge = new Integer(j);
+	match = false;
+	if ((j >= SOCBoard.MINEDGE) && (j <= SOCBoard.MAXEDGE) &&
+	    (ourPlayerData.isLegalRoad(j))) {
+	  for (Enumeration ev = visited.elements();
+	       ev.hasMoreElements(); ) {
+	    Integer vis = (Integer)ev.nextElement();
+	    if (vis.equals(edge)) {
+	      match = true;
+	      break;
+	    }
+	  }	
+	  if (!match) {
+	    Vector newVis = (Vector)visited.clone();
+	    newVis.addElement(edge);
+	    // node coord = edge coord + 0x01
+	    j += 0x01;
+	    pending.push(new Pair(new NodeLenVis(j, len+1, newVis), dataPair));
+	    pathEnd = false;
+	  }
+	}
+      }		
+      
+      if (pathEnd) {
+	if (len > longest) {
+	  longest = len;
+	  numRoads = curNode.len - pathLength;
+	  bestPathNode = dataPair;
+	} else if ((len == longest) &&
+		   (curNode.len < numRoads)) {
+	  numRoads = curNode.len - pathLength;
+	  bestPathNode = dataPair;
+	}
+      }
+    }
+    if ((longest > lrLength) &&
+	(bestPathNode != null)) {
+      D.ebugPrintln("Converting nodes to road coords.");
+      //
+      // return the path in a stack with the last road on top
+      //
+      //
+      // convert pairs of node coords to road coords
+      //
+      Stack temp = new Stack();
+      SOCPossibleRoad posRoad;
+      int coordA, coordB, test;
+      Pair cur, parent;
+      cur = bestPathNode;
+      parent = (Pair)bestPathNode.getB();
+      while (parent != null) {
+	coordA = ((NodeLenVis)cur.getA()).node;
+	coordB = ((NodeLenVis)parent.getA()).node;
+	test = coordA - coordB;
+	if (test == 0x11) {
+	  // it is a '\' road
+	  D.ebugPrintln(game.getBoard().nodeCoordToString(coordB));
+	  posRoad = new SOCPossibleRoad(ourPlayerData, coordB, new Vector());
+	} else if (test == -0x11) {
+	  // it is a '/' road
+	  D.ebugPrintln(game.getBoard().nodeCoordToString(coordA));
+	  posRoad = new SOCPossibleRoad(ourPlayerData, coordA, new Vector());
+	} else if (test == 0x0F) {
+	  // it is a '|' road for an A node
+	  D.ebugPrintln(game.getBoard().nodeCoordToString((coordA - 0x10)));
+	  posRoad = new SOCPossibleRoad(ourPlayerData, (coordA - 0x10), new Vector());
+	} else {
+	  // it is a '|' road for a Y node
+	  D.ebugPrintln(game.getBoard().nodeCoordToString((coordA - 0x01)));
+	  posRoad = new SOCPossibleRoad(ourPlayerData, (coordA - 0x01), new Vector());
+	}
+	temp.push(posRoad);
+	cur = parent;
+	parent = (Pair)parent.getB();
+      }
+      //
+      // reverse the order of the roads so that the last one is on top
+      //
+      Stack path = new Stack();
+      while (!temp.empty()) {
+	path.push(temp.pop());
+      }
+      return path;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * smart game strategy
+   * use WGETA to determine best move
+   *
+   * @param buildingETAs  the etas for building something
+   */
+  protected void smartGameStrategy(int[] buildingETAs)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 79. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1207-1217 into a separate method.